docs/development/design/fsm.md

```diff
@@ -0,0 +1,57 @@
+# Nexd State Machine
+
+> [Issue #885](https://github.com/nexodus-io/nexodus/issues/885)
+
+## Summary
+
+State management in nexd is becoming more complicated.
+We need a mechanism to allow us to simplify that logic and not rely on crash/restart to retrigger things.
+
+## Proposal
+
+I propose that we implement a Finite State Machine inside of nexd that operates as follows:
+
+```mermaid
+---
+title: Nexd State Machine
+---
+stateDiagram-v2
+    [*] --> IDLE
+    IDLE --> AUTH: "login"
+    AUTH --> READY: "login_success"
+    AUTH --> IDLE: "login_failed"
+    READY --> PEERING: "start"
+    READY --> IDLE: "logout"
+    PEERING --> READY: "stop"
+    PEERING --> READY: "peering_error"
+    PEERING --> IDLE: "authentication_error"
+    PEERING --> IDLE: "logout"
+```
+
+nexd may be in one of four states:
+
+1. IDLE: No configuration has been provided
+2. AUTH: Performing authentication, either interactively or via token
+3. READY: Authenticated and ready to peer
+4. PEERING: Actively Peering
+
+To provide information to the user about the current state, a `nexctl nexd status` command will be added. This can inform the user of the current state and what to do next.
+
+To transition from IDLE to AUTH the "login" transition must occur.
+
+The trigger for this will be a CLI command: `nexctl nexd login`
+
+Once in the "AUTH" state, credentials will be taken from the user - or the device-auth flow will occur. In event of successful login the "login_success" transition occurs to bring nexd into the READY state. In case of "login_failed" we go back to IDLE.
+
+When in the READY state we will immediately issue a transition "start" into the PEERING state to preserve the existing behavior. The commands `nexctl nexd start` and `nexctl nexd stop` will be added to manually issue these transitions so users can choose to start/stop peering as necessary.
```

Comment:
I think the `stop` state is pretty risky. That basically means that the device is connected to the control plane, but it's not controlled by control plane in the `stop` state. So any changes in the peering/policies (which possibly doesn't allow that device to talk to some of the existing devices in-network or out-of-network anymore) won't be enforced, and that's not a good state. I think we should keep `login` and `logout` commands and ditch the `start` and `stop`. Login means you are onboarded and logout means you are offboarded.

Reply:
> That basically means that the device is connected to the control plane, but it's not controlled by control plane in the stop state. 

No that's not the intention. "stop" is not the state. READY is the state.
In READY you are "ready" to connect to the control plane and start peering - you are not connected to the control plane and there is no dataplane setup.

> I think we should keep login and logout commands and ditch the start and stop. Login means you are onboarded and logout means you are offboarded.

Think of the following user story though:

As a Nexodus user, I want to pause my mesh VPN connection to perform a task, then reconnect afterwards without having to reauthenticate with the control plane - i.e you're working on a local VM whose IP address range overlaps with that on the mesh.

Forcing a user through the login/logout flow is suboptimal IMO.

Reply:
>No that's not the intention. "stop" is not the state. READY is the state.
In READY you are "ready" to connect to the control plane and start peering - you are not connected to the control plane and there is no dataplane setup.

Okay let me try to clarify my picture from control plan perspective. 
When device moves from IDLE->AUTH state, does control plane see the device ? 
When device moves from PEERING -> READY as a result of `stop` command, does that means the agent cleanup it's peering configuration and disconnect from all the peers?

>As a Nexodus user, I want to pause my mesh VPN connection to perform a task, then reconnect afterwards without having to reauthenticate with the control plane - i.e you're working on a local VM whose IP address range overlaps with that on the mesh.

As far as `stop` means that it clears it's local peering state and disconnect from others in the mesh network, and control plane sees that device is gone and update the other devices in the network about that, its seems reasonable to me. 

But if "pause" means it still maintain the connection with other peers, where other peers are still connected and sending packets in, while it edits it's local state, i am not very convenience it's a practical use case and seems like a bad violation of zero trust. 

Just to make sure i am putting my point across clearly, just throwing my local state to action map here 

login --> control plan detect the device, no peering information sent to new and existing device, token stored by new device

start --> control plan send initial peering information to the new device, also sends updated peering information to existing devices in network,

stop -> Agent remove the device from control plan, clear local peering information, control plan sends updated peering information to existing devices in the network

logout -> Control plan invalidate login token, next removes locally stored api token.

Reply:
Added better descriptions to the states to help clarify control/data plane connections.